include/wil/cppwinrt_authoring.h

```diff
@@ -0,0 +1,219 @@
+    >
+    struct notify_property_changed_base {
+    public:
+        using Type = T;
+        auto PropertyChanged(Xaml_Data_PropertyChangedEventHandler const& value) {
+            return m_propertyChanged.add(value);
+        }
+        void PropertyChanged(winrt::event_token const& token) {
+            m_propertyChanged.remove(token);
+        }
+        Type& self() {
+            return *static_cast<Type*>(this);
+        }
+
+        /**
+         * @brief Raises a property change notification event
+         * @param name The name of the property
+         * @return
+         * @details Usage example\n
+         * C++
+         * @code
+         * void MyPage::DoSomething() {
+         *  // modify MyInt
+         *  // MyInt = ...
+         *
+         *  // now send a notification to update the bound UI elements
+         *  RaisePropertyChanged(L"MyInt");
+         * }
+         * @endcode
+        */
+        auto RaisePropertyChanged(std::wstring_view name) {
+            return m_propertyChanged(self(), Xaml_Data_PropertyChangedEventHandler{ name });
+        }
+    protected:
+        winrt::event<Xaml_Data_PropertyChangedEventHandler> m_propertyChanged;
+    };
+
+    /**
+     * @brief Implements a property type with notifications
+     * @tparam T the property type
+     * @details Use the #INIT_NOTIFY_PROPERTY macro to initialize this property in your class constructor. This will set up the right property name, and bind it to the `notify_property_changed_base` implementation.
+    */
+    template<typename T,
+        typename Xaml_Data_PropertyChangedEventHandler = winrt::Windows::UI::Xaml::Data::PropertyChangedEventHandler,
+        typename Xaml_Data_PropertyChangedEventArgs = winrt::Windows::UI::Xaml::Data::PropertyChangedEventArgs>
+    struct single_threaded_notifying_property : single_threaded_rw_property<T> {
+        using Type = T;
+
+        using single_threaded_rw_property<T>::operator();
+
+        void operator()(const T& value) {
+            if (value != this->m_value) {
+                single_threaded_rw_property<T>::operator()(value);
+                if (m_npc) {
+                    (*m_npc)(m_sender, Xaml_Data_PropertyChangedEventArgs{ m_name });
+                }
+            }
+        }
+        template<typename... TArgs>
+        single_threaded_notifying_property(
+            winrt::event<Xaml_Data_PropertyChangedEventHandler>* npc,
+            winrt::Windows::Foundation::IInspectable sender,
+            std::wstring_view name,
+            TArgs&&... args) :
+            single_threaded_rw_property<T>(std::forward<TArgs...>(args)...) {
+            m_name = name;
+            m_npc = npc;
+            m_sender = sender;
+        }
+
+        single_threaded_notifying_property(const single_threaded_notifying_property&) = default;
+        single_threaded_notifying_property(single_threaded_notifying_property&&) = default;
+    private:
+        std::wstring m_name;

```

Comment:
Is there a way to make this a method instead of storage on the instance?

Or should this store the `std::wstring_view` so it's not making a copy here, since it's a preprocessor-generated string?

Reply:
if we can guarantee RTTI, then we could rely on it, but that seems like a niche case since most usages turn it off. using wstring_view seems like it might work though.